addons/hr_holidays/models/hr_leave_allocation.py

```diff
@@ -825,7 +825,7 @@ def action_refuse(self):
         if any(allocation.state not in ['confirm', 'validate'] for allocation in self):
             raise UserError(_('Allocation request must be confirmed or validated in order to refuse it.'))
 
-        days_per_allocation = self.employee_id._get_consumed_leaves(self.holiday_status_id)[0]
+        days_per_allocation = self.employee_id.with_context(allocation_refuse_check=True)._get_consumed_leaves(self.holiday_status_id)[0]
 
         for allocation in self:
             days_taken = days_per_allocation[allocation.employee_id][allocation.holiday_status_id][allocation]['virtual_leaves_taken']
```

Comment:
Did you check to simply remove the condition `if days_taken > 0`? As I read the way `write` is executed, we check the excess days before and after the write, so maybe that the issues with incoherences will be caught by it without the need for this extra condition